Music and playlists

samrocketman opened the pull request:
* This PR is a revival of #7538.
* Authorship from prior code review is preserved.

Original author: @RisingLeaf 
Co-authored-by: @Hurleveur
Co-authored-by: @tibetiroka

Feature Summary
---------------

Adds the ability to define event based music and sounds.

- Music has a dedicated volume slider in settings.  The slider is hidden by default if there are no playlists.
- Music or sounds can play during conversations via setting conditions.  This feature does not need to be limited to music.  For example, a bar scene with conversation noise and laughter can play during a conversation.
- Music/audio is defined as tracks.
- Tracks are organized into playlists.
- Playlists are triggered (start playing) by [player conditions][conditions] and can be [filtered by location][location].
- Tracks of silence are supported.

> Please keep in mind this feature lays the ground work for future enhancements to music.  Aside from game breaking bugs I would like this feature to be accepted as-is with the intent to open many followup PRs with new enhancements to the ES music system.  This is the first music system beyond just looping a single audio file defined in a single system or single planet.

[conditions]: https://github.com/endless-sky/endless-sky/wiki/Player-Conditions
[location]: https://github.com/endless-sky/endless-sky/wiki/CreatingMissions#filters

# Related issues

- closes https://github.com/endless-sky/endless-sky/issues/6293
- https://github.com/endless-sky/endless-sky/issues/6596 to be addressed in future PR.  This PR does work to trigger music during a conversation but I don't think it is satisfactory because it will be ignored if music is already playing.
- https://github.com/endless-sky/endless-sky/issues/6479
- https://github.com/endless-sky/endless-sky/issues/4778
- https://github.com/endless-sky/endless-sky/issues/3594
- https://github.com/endless-sky/endless-sky/issues/233

# New Playlist and Track data syntax

## Track Syntax

```html
track <name>
	combat <sound>
	idle <sound>
	landed <sound>
	[volume <double>]
	[wait <number of seconds>]
```

`track` followed by the `name` of the track is how a track can be defined.  This can play different sounds.

* `combat` sound is played during combat.
* `idle` sound is played while player is idle in space.
* `landed` sound is played when landed on a planet.
* `volume` (default: `0`; range between `-1` and `1`) When a track plays you can specify that it be played with an increased or decreased `volume` modified from the player preference volume slider.  For example, `volume 0.1` will increase the playback volume.
* `wait` after playing a track.  A track can force itself to `wait` for a specified number of seconds after it finishes playing.

Example track

```ruby
track "5 minutes of silence"
	combat ""
	idle ""
	landed ""
	wait 300
```

## Playlist Syntax

```html
playlist <name>
	to play
		<condition> <comp> <value>
		(has | not) <condition>
		never
		(and | or)
			...
	location
		[<modifier>] planet <name>...
			<name>...
		[<modifier>] system <name>...
			<name>...
		[<modifier>] government <name>...
			<name>...
		[<modifier>] attributes <name>...
			<name>...
		[<modifier>] outfits <name>...
			<name>...
		[<modifier>] category <name>...
		[<modifier>] near <system> [[<min>] <max>]
		[<modifier>] distance [<min>] <max>
		neighbor
			...
		not
			...
	[priority <number 0 or greater>]
	[weight <number 1 or greater>]
	tracks [(random | linear | pick)]
		<track name> [<weight>]
		...
```

`playlist` followed by the `name` of the playlist.  It has one required option to be a valid playlist that plays sound.

* `tracks` followed by a list of tracks.
  * Each track entry in the `tracks` list is the name of the track _optionally followed by_ a track `weight` (likelihood that track is picked initially or during each shuffle at `random`).  If not specified, the `default weight` is `1` for tracks.

And `playlist` has additional optional options.

* `to play` is the [player conditions][conditions] in which tracks have a chance to play.  For example, play during active missions or a condition set during conversation.  If not specified, then it always assumes it should play given any condition.
* `location` uses a [location filter][location] which limits where the play list is allowed to play.  If no location is specified, then every location is valid.
* `priority` (`0` or higher) sets the playlist priority when multiple playlists match `to play` and `location`.  Higher priority is always played.  `0` is lowest priority.
* `weight` (`1` or higher) if two playlists have the same priority, then weight is a factor in chance for that given playlist to be selected in a list of playlists matching `to play`, `location`, and `priority`.  Increased weight means the playlist has a higher likelihood of being played.
* `tracks [(random | linear | pick)]` continued... There are three ways track progression can be selected.  If the a progression is not provided, then the default is `linear`.
  * `random` will shuffle tracks and pick tracks according to their `weight` (likely chance of being picked).
  * `linear` tracks will play linearly.  An initial track is selected based on weights at random.  From the selected track onward the track progression will be linear in the list of defined tracks.
  * `pick` will pick a single track to play.  The playlist will keep repeating the track until playlist conditions no longer match.  A track is selected at random with a higher `weight` increasing the chance a track will be picked.

Example playlist, which has varying silent periods to interrupt other musical playlists.

```ruby
playlist "A silent period"
	to play
		0 == 0 #always play
	tracks pick
		"5 minutes of silence"
```

# Detailed Example

Add a playlist to the game that is only played in certain locations and if certain conditions are met.

Example:

```ruby
track "test2"
	volume 0.1 #add 0.1 to volume
	idle "music/mainscreen" #idle track, played during normal gameplay
	combat "music/mainscreen" #combat track, played during combat
	landed "music/mainscreen" #landed track, played while landed

track "test1"
	idle "music/haispace"
	combat "music/haispace"
	landed "music/haispace"

playlist "test"
	to play #conditions for playing this playlist
		has "Hai Intro: done"
	location #locationfilter for where to play this
		not government "Hai"
	priority 1 #the priority, higher priority playlist will be chosen over lower priority playlists
	weight 10 #the weight of a playlist inside of its priority layer, higher weight means more probable to be chosen
	tracks "linear"
		"test1" 20
		"test2" 20
```

Example Explanation
--------------------

`track NAME` holds the following child keys.

* `NAME` the name of the track
* `volume` (default 0): a value that should be chosen between -1 and 1 that gets added to the current volume of the music. (use this with small values, otherwise users who want dim music will be annoyed.)
* Three categories of music associated with this track. The tracks will be played when the player is in their state. If one is not given, nothing will be played in that state.
  * `idle`
  * `combat`
  * `landed`

`playlist NAME` holds the following child keys.

* `NAME` the name of the playlist.
* `to play` conditions that determine if this playlist can be played.
* A `location` filter for where this playlist will be played.
* A `priority`, higher priorities will always play over lower priorities.
* A `weight`, playlists with higher weight are more likely to be played inside the same priority.
* `tracks`
  * A progression style: `linear`, `random`, or `pick` (pick means that one track will be picked and repeated until the playlist is switched)
  * Various tracks with weights to them, weights only matter for pick or random progression style.

This would play either track one or track two when you are not in Hai government zone.

Testing Done
------------

- [x] Do the tutorial while music plays.
- [x] The music should not play if the tutorial is declined.
- [x] The music should stop playing if the tutorial is completed.
- [x] Music should be seamlessly playing when jumping across systems and landed on planets.
- [x] Music should seamlessly play when the game is paused.
- [x] If the music stops mid-tutorial it should not play again.
- [x] The volume slider should work on the music.  Verify that volume for ships and music is independent.
- [x] Check the High DPI volume slider operates okay.
- [x] Music volume slider should be hidden if there's no music playlists.
- [x] Playlist should fade out instead of abruptly stopping.
- [x] Start the game with no saves (fresh player start)
- [x] Verified `5 minutes of silence` track waits for 5 minutes with no music playing in a playlist of music.

Test data: [sounds-of-endless-sky plugin][soes] playing music only during the first-time player intro tutorial series.

[soes]: https://github.com/samrocketman/sounds-of-endless-sky/blob/97ee8eba75ea06d39a778dec404de8fc4c82ba23/data/republic.txt

Checklists
---------

- [x] Update with upstream master including CMake.
- [x] Do some testing.
- [x] Remove from draft when all checklists are satisfied.

### Artwork Checklist

 - [x] I updated the copyright attributions, or decline to claim copyright of any assets produced or modified
 - [x] I created a PR to the [endless-sky-assets repo](https://github.com/endless-sky/endless-sky-assets) with the necessary image, blend, and texture assets: https://github.com/endless-sky/endless-sky-assets/pull/127
 - [x] I created a PR to the [endless-sky-high-dpi repo](https://github.com/endless-sky/endless-sky-high-dpi) with the `@2x` versions of these art assets: https://github.com/endless-sky/endless-sky-high-dpi/pull/352

samrocketman commented:
@RisingLeaf I need to add copyrights.  Can you tell me how you want your artwork copyrighted?

RisingLeaf commented:
> @RisingLeaf I need to add copyrights. Can you tell me how you want your artwork copyrighted?

do as you please actually, but I think it needs to be copyrighted?

RisingLeaf commented:
> Copyright: RisingLeaf (https://github.com/RisingLeaf)
> License: CC-BY-SA-4.0

this is the one I usually use

samrocketman commented:
The following throws an exception and crashes the game

```
track "Republic Space"
	idle "music/ByAnthrophantasmus/Republic Space"
	combat "music/ByAnthrophantasmus/Republic Space"
	landing "music/ByAnthrophantasmus/Republic Space"
playlist "Republic Intro"
	location
		government Republic
	tracks linear
		"Republic Space" 1
```

which is thrown by

```
	// All weights must be >= 1.
	if(weight < 1)
		throw std::invalid_argument("Invalid weight inserted into weighted list. Weights must be >= 1.");
```
in `source/WeightedList.h` line 164

samrocketman commented:
I can't seem to get music to play with the following data file.

```ruby
track "Republic Space"
        idle "music/ByAnthrophantasmus/Republic Space"
        combat "music/ByAnthrophantasmus/Republic Space"
        landing "music/ByAnthrophantasmus/Republic Space"

playlist "Republic Intro"
        to play
                has "Intro [0]: offered"
        location
                government Republic
        weight 10
        priority 1
        tracks linear
                "Republic Space" 10
```

Using the default music definition I can get it to play in 1 system.

```ruby
system Rutilicus
       music "music/ByAnthrophantasmus/Republic Space"
planet "New Boston"
       music "music/ByAnthrophantasmus/Republic Space"
```

samrocketman commented on source/Audio.cpp:
```diff
@@ -255,6 +286,90 @@ void Audio::Update(const Point &listenerPosition)
+	}
+	else
+	{
+		if(player.GetSystem()->MusicName().empty())
```
This could be related as to why music may not be playing but there are other issues too.

samrocketman replied:
Music interrupting while jumping across systems is likely a separate bug.

samrocketman replied:
Fixed

samrocketman commented:
Playback works well now.  I'm going to do a few extra tests and then I will take it out of draft this evening.

samrocketman commented:
Thanks @RisingLeaf for some additional support

samrocketman commented:
This is now ready for review.  All found bugs resolved and testing checklist completed.

samrocketman commented:
When this is merged I plan to write up a content creator's guide in the wiki.

samrocketman commented on source/Playlist.h:
```diff
@@ -0,0 +1,67 @@
+	LocationFilter location;
+
+	int priority = 0;
+	int weight = 0;
```
```suggestion
	int weight = 2;
```
Set the default weight to 2 so that a content creator is not forced to choose a weight.  Default to the lowest weight.

samrocketman replied:
Fixed; now default is 1.

samrocketman commented on source/Playlist.cpp:
```diff
@@ -0,0 +1,134 @@
+				if(grand.Size() >= 2)
+					tracks.emplace_back(grand.Value(1), GameData::Tracks().Get(grand.Token(0)));
+				else
+					tracks.emplace_back(10, GameData::Tracks().Get(grand.Token(0)));
```
The default track weight is `10` if the content creator does not specify a track weight.  I asked RisingLeaf about this.  We may want to change the default weight to `2`.

Also, the game crashes with an exception thrown any time a content creator specifies a `weight` less than `2`.  We should probably throw a syntax warning and ignore the weight value instead of crashing.

samrocketman replied:
Fixed; now default is 1

samrocketman commented on source/Playlist.cpp:
```diff
@@ -0,0 +1,134 @@
+	else if(progressionStyle == "pick")
+		return currentTrack;
+
+	// Asuming the progression style is "random".
+	return tracks.Get();
```
This assumes the content creator chose `tracks random` as the progression style.  We'll probably want to throw a warning in `errors.txt` if a content creator chose anything except for `tracks (random | linear | pick)`.  To warn the creator they chose something we didn't expect but we're going to default to a `random` shuffle.

samrocketman replied:
Fixed

samrocketman commented:
This PR is ready for merge.  Bugs have been squashed.  Enhancements have been identified but they're out of scope for this pull requests.  I won't be doing further enhancements until I can make followup PRs.

This is necessary because I adopted someone else's PR and it is now in good shape for how I'd like to see music considering I'm leading its coordination.

samrocketman commented:
Added related issues section to description.  Identified a reported issue to close

samrocketman commented:
Play testing for this PR should be done before merging; it has been 6 months since this was last looked at by me and a lot has changed in ES.  I'll want to make sure it remains polished like it was when it was introduced.

samrocketman reviewed (commented):
I've discovered a playback issue; playlist relies on WeightedList and `tracks->Get()` is a shuffle based on weight 100% of the time so `linear` isn't actually `linear`.  This is a bug which needs to be fixed before merged.

i.e. `tracks linear` needs to actually be linear

samrocketman commented:
@warp-core this has been updated with the latest master, I've tested the changes and resolved any issues found during testing.

This is ready for review.

Hurleveur commented on source/Engine.cpp:
```diff
@@ -554,6 +556,16 @@ void Engine::Step(bool isActive)
+		Track::GameState state = Track::GameState::IDLE;
+		// If there are hostile ships in the system, set the state to COMBAT.
+		if(hadHostiles)
+			state = Track::GameState::COMBAT;
```
I'm wondering, will this fade in and out or will it give boss music for a sparrow and then stop suddenly?

samrocketman replied:
Its transitions need work; this is a feature gap which should be fixed in a future enhancement

Hurleveur replied:
fair, it probably isnt as simple as adding a time before it activates and then a cooldown before it can go off

Hurleveur commented on source/GameData.cpp:
```diff
@@ -652,6 +652,21 @@ const Set<Planet> &GameData::Planets()
+
+bool GameData::HasPlaylists()
+{
+	return objects.playlists.empty();
```
```suggestion
	return !objects.playlists.empty();
```
am I confused or is this a fix?

samrocketman replied:
May be a fox but I need to look through the code to verify what else needs to change.  I play tested it and from a data file playtime perspective the code works so likely changing this means I'll need to invert it in other places

samrocketman replied:
I still need to look through and verify this.

It is needed for unreleased features.  Overall, this feature is a starting point and not the end of the maturity of the endless sky music engine.  Some of the implementation was made here anticipating its need based on implementation of other sub-systems within ES code base and how I think content creators would use music as part of writing stories.

RisingLeaf replied:
function has been removed

Hurleveur commented on source/Music.h:
```diff
@@ -47,6 +47,9 @@ class Music {
 	const std::vector<int16_t> &NextChunk();
 
+	bool IsFinished() const;
+	void Finish();
```
we using this one?
should we be using it?

samrocketman replied:
It does not appear to be used but it makes sense to have a setter where you have a getter to short-cut it.  This engine support is more "initial support for music" by RisingLeaf and I polished it a little bit fixing bugs where I found them.  I think it should probably stay available as we expand on music support with new features.

For example, event-based music starting should stop existing music if it has a higher priority.

Hurleveur commented on source/Track.cpp:
```diff
@@ -0,0 +1,103 @@
+			break;
+		default:
+			return idleTitle;
+			break;
```
mb make the default one also the IDLE one by moving it lower?
does that actually matter performance wise? I'm not sure but I think the switch only looks at it once anyway?

samrocketman replied:
This could be an enhancement; I would like to avoid changing the original code as much as possible beyond getting it into a workable state.

If you do experience a performance issue in testing it should be addressed before merging

Hurleveur replied:
I think you misunderstood, I simply meant we could merge default and IDLE
and I was asking if it wouldnt be a bad idea perf wise

samrocketman replied:
Considering GameState only has three states I don't see how `default` would ever get triggered; does it make sense to just delete it entirely?

Hurleveur replied:
I mean yeah sure

samrocketman replied:
Worth noting I didn't write most of the code; just adopted it and am cleaning it up.  Is there a scenario where it is ever possible for one of the three game states to not be one of the three game states passed as an option?

RisingLeaf replied:
i also just noticed that we dont need the break; in any line

Hurleveur reviewed (commented):
Lets get this PR rolling
Looking good, I'll try to test it at some point

RisingLeaf commented on source/Playlist.cpp:
```diff
@@ -0,0 +1,146 @@
+				progressionStyle = child.Token(1);
+			else
+			{
+				if(child.Size() >= 2)
```
we are using child.Size >= 2, might as well create a variable called hasChildren

Amazinite replied:
This is the number of tokens in the line, not the number of children of the node. In any case, I made a suggestion to add a hasValue variable.

RisingLeaf commented on source/Playlist.cpp:
```diff
@@ -0,0 +1,146 @@
+			weight = max<int>(1, child.Value(1));
+		else if(child.Token(0) == "tracks")
+		{
+			std::vector<std::string> progressionStyles = {"random", "linear", "pick"};
```
lets make this a const in the unamed namespace

RisingLeaf commented on source/Playlist.cpp:
```diff
@@ -0,0 +1,146 @@
+
+bool Playlist::MatchingConditions(PlayerInfo &player) const
+{
+	if(player.GetPlanet())
+		if(!location.Matches(player.GetPlanet()))
+			return false;
+	return toPlay.Test(player.Conditions()) && location.Matches(player.GetSystem());
```
we can make this a one liner

samrocketman replied:
I'll look at cleaning up this logic.

RisingLeaf replied:
cleaned

RisingLeaf commented on source/Playlist.h:
```diff
@@ -0,0 +1,67 @@
+	ConditionSet toPlay;
+	LocationFilter location;
+
+	int priority = 0;
+	int weight = 1;
```
should these be unsigned as a fail safe?

samrocketman replied:
I don't know the right answer to this.  Can you elaborate why this is unsafe as is?

RisingLeaf replied:
they should never be less than 0, changing

RisingLeaf commented on source/Track.cpp:
```diff
@@ -0,0 +1,103 @@
+#include <iostream>
+
+using namespace std;
+
```
```suggestion



```

RisingLeaf commented on source/Track.cpp:
```diff
@@ -0,0 +1,103 @@
+		else if(child.Token(0) == "landed" && child.Size() >= 2)
+			landedTitle = child.Token(1);
+		else if(child.Token(0) == "wait" && child.Size() >= 2)
+			wait = max<int>(0, child.Value(1));
```
lets throw a warning if it is less than zero?

Hurleveur replied:
I dont think the rest of the code does that?

RisingLeaf replied:
hm true, ~maybe it should though~

Hurleveur replied:
~~yk as well as I do that should be in another PR then~~

RisingLeaf commented on source/Track.h:
```diff
@@ -0,0 +1,63 @@
+// Class to store a track of music that can be used in a playlist.
+class Track {
+public:
+	enum class GameState
+	{
+		IDLE,
+		COMBAT,
+		LANDED
+	};
```
```suggestion
	enum class GameState : int_fast8_t
	{
		IDLE = 0,
		COMBAT,
		LANDED
	};
```

RisingLeaf reviewed (commented):
reviewing my own code XD

samrocketman commented:
I plan to wait for a core dev to review and then I'll address everyone's feedback in one go.

Hurleveur commented:
The way I see it we reviewers are there to make dev life easier by taking care of the small and obvious stuff

samrocketman commented:
> The way I see it we reviewers are there to make dev life easier by taking care of the small and obvious stuff

I've had multiple interactions where feedback would conflict which complicates getting it merged.

This PR age is quite old.  I would agree if this were a new PR but I am adopting someones older code trying to minimize changes to fixes only.  I'm not looking to add features or enhancements.

My PRs don't seem to be priority which is why I have essentially stopped all development on ES. So unless devs meet me half way on existing PRs like this one I will continue to be on hiatus. I don't feel it is a good use of my time if contributions are ignored from the standpoint of the game getting extended (i.e. PRs reviewed and merged in a timely manner).

I do appreciate your reviews; I do plan to incorporate your feedback in this or future PRs.

Amazinite commented on source/Playlist.cpp:
```diff
@@ -0,0 +1,146 @@
+using namespace std;
+
+namespace {
+	const Track *currentTrack = nullptr;
```
Why store this in the namespace instead of in the header?

RisingLeaf replied:
as a static?

samrocketman replied:
As of right now it's in the namespace since this is the only place that would reference it.

Do you want this moved into header to proceed?

Amazinite commented on source/Playlist.cpp:
```diff
@@ -0,0 +1,146 @@
+
+void Playlist::Load(const DataNode &node)
+{
+	// All tracks need a name.
+	if(node.Size() < 2)
+	{
+		node.PrintTrace("Error: No name specified for playlist:");
+		return;
+	}
+
```
Already handled by UniverseObjects.
```suggestion
```

Amazinite commented on source/Playlist.cpp:
```diff
@@ -0,0 +1,146 @@
+
+	for(const DataNode &child : node)
+	{
+		if(child.Token(0) == "to")
```
Ditto from Track.
```suggestion
		const string &key = child.Token(0);
		bool hasValue = child.Size() >= 2;
		if(key == "to")
```

Amazinite commented on source/Playlist.cpp:
```diff
@@ -0,0 +1,146 @@
+		else if(child.Token(0) == "location")
+			location.Load(child);
+		else if(child.Token(0) == "priority" && child.Size() >= 2)
+			priority = child.Value(1);
```
You said 0 is the lowest priority in the PR description, yet this is allowed to go negative.

samrocketman replied:
I'll add a value check and throw a warning if less than zero to warn user zero was set instead.

Amazinite commented on source/Playlist.cpp:
```diff
@@ -0,0 +1,146 @@
+		{
+			std::vector<std::string> progressionStyles = {"random", "linear", "pick"};
+			bool validProgressionStyle = (child.Size() >= 2) ?
+					count(progressionStyles.begin(), progressionStyles.end(), child.Token(1)) :
```
Why not make progressionStyles a set? That way you can just do set.count(token).

Amazinite commented on source/Playlist.cpp:
```diff
@@ -0,0 +1,146 @@
+{
+	if(progressionStyle == "linear")
+	{
+		const Track * tmpTrack = currentTrack;
```
```suggestion
		const Track *tmpTrack = currentTrack;
```

Amazinite commented on source/Engine.cpp:
```diff
@@ -64,6 +64,8 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 #include "System.h"
 #include "SystemEntry.h"
 #include "Test.h"
+#include "TestContext.h"
```
This include seems to have snuck its way in from a merge conflict.

Amazinite commented on source/PlayerInfo.cpp:
```diff
@@ -1484,7 +1484,8 @@ void PlayerInfo::Land(UI *ui)
 	{
 		Audio::Play(Audio::Get("landing"));
-		Audio::PlayMusic(planet->MusicName());
+		if(!planet->MusicName().empty())
+			Audio::PlayMusic(planet->MusicName());
```
Is this not made redundant by the first part of Audio::UpdateMusic?

RisingLeaf replied:
yes

Amazinite commented:
Feature-wise, this is looking very nice for what I'd imagine out of a music system for the game. One thing that seems missing at the moment though is the ability to play music at the main menu.

Also, while I see that we can set an exact number of seconds after a track has ended to wait before another track plays, I'm wondering if there's some better method of having pauses between tracks. I see the following potential issues with the current setup:
1. If you have combat music, you presumably wouldn't want to have to wait for a timer to tick down since the last track played. Combat music is something you want to have pick up in the moment.
2. If we decide that we want lots of waits between songs (being a bit more like Minecraft where the music isn't constant), we need to add long waits to every track we have.
3. It seems to me like we can't have a wait that occurs before a track plays. That means whenever you boot up the game, a new track is going to play the first opportunity it gets, rather than being able to have a pause after starting the game before the first track can start playing.
4. We can't have random wait times between tracks, only an exact number of seconds after a certain track that is the same every time. (Rather minor and likely easy to fix.)

I'm wondering if the wait time between tracks would make more sense to be attached to playlists rather than attaching it to individual tracks.

RisingLeaf commented on source/PreferencesPanel.cpp:
```diff
@@ -206,6 +209,11 @@ bool PreferencesPanel::Click(int x, int y, int clicks)
 	}
+	else if(x >= 309.5 && x < 340.5 && y >= -220 && y < 70)
+	{
+		Audio::SetMusicVolume((20 - y) / 200.);
```
Yes, but we would a need a small music for that?

samrocketman replied:
I think in general this is a good idea; would it be acceptable to open a followup PR for this?  I don't know the complexity of the ask currently without trying to do it.

Probably okay to reuse the same sound.

RisingLeaf replied:
for now using warder sound

RisingLeaf commented on source/Track.cpp:
```diff
@@ -0,0 +1,97 @@
+	{
+		const string &key = child.Token(0);
+		bool hasValue = child.Size() >= 2;
+		if(key  == "volume" && hasValue)
```
```suggestion
		if(key == "volume" && hasValue)
```

Terin commented on source/Track.cpp:
```diff
@@ -0,0 +1,97 @@
+			idleTitle = child.Token(1);
+		else if(child.Token(0) == "combat" && child.Size() >= 2)
+			combatTitle = child.Token(1);
+		else if(child.Token(0) == "landed" && child.Size() >= 2)
```
The PR documentation has this listed as `landing`, while the original design doc has it as `landed` along with the code here. Should make it consistent one way or the other.

samrocketman replied:
Should probably match mission syntax

Hurleveur replied:
landing might convey the idea that it only happens once when you first land and then not anymore (which would be fine ig but isnt whats happening?)

samrocketman replied:
I updated the PR description to `landed`.  If it changes in code I guess let me know and I can update it again.

samrocketman replied:
Resolving since I think this was addressed by updating PR description.

petervdmeer commented:
"Waiting on OP"; I see a number of not-yet-addressed open review comments.

samrocketman commented:
> "Waiting on OP"; I see a number of not-yet-addressed open review comments.

The purpose of this PR was to design the music format of the data files and not have a perfect feature.

Any followup enhancements can be provided by additional pull requests.  Even when merged this feature will not be in use.  It is a significant improvement over the current implementation.

Perfect is the enemy of good.

petervdmeer commented:
> > "Waiting on OP"; I see a number of not-yet-addressed open review comments.
> 
> The purpose of this PR was to design the music format of the data files and not have a perfect feature.

Fine for me.

> Any followup enhancements can be provided by additional pull requests.

Fine for me as well, except that when somebody takes the time to write a relevant comment, then I expect the author of a PR to at least reply to such comment. Communication is very important, and a reply like "not in this PR, I will do this in a follow-up feature.", or even a reply like "not in this PR, I consider this out of scope and I'm not planning to pick this up at all" is fine for me (except when the comment is really relevant, then I just expect the comment to be handled correctly with a fix).

I'm not okay with comments that don't get a reply; your time is very valuable, but equally so is time of the persons that make an effort to do a good review your PR. (And we actually have more people creating PRs than that we have reviewers, so good reviews are in short supply. Don't waste them.) 

> Even when merged this feature will not be in use. It is a significant improvement over the current implementation.

I'm sure it is, but this PR currently has a merge-conflict, and I see very sensible comments that should be 10 minutes to fix and didn't get a fix and marked-as-resolved.

> Perfect is the enemy of good.

You were complaining about the devs wasting your time by not reviewing your PRs fast enough (and you do have a point about reviews taking very very long), but when this dev looks at this PR today, then this PR itself is not ready for merge.

If you want your PRs to be merged in an open source project like this, then you need to spend quite a bit of effort to get your PRs ready to merge and to keep them ready for merge. Projects like this run on volunteers, and I happen to have some time available today to review and merge PRs. But there are 150 open PRs that have an author that hopes for a (timely) review (and possibly merge) on their PR, so I feel it is unfair if I spend the limited time I have for this project on PRs that have open comments or that are not ready for merge.

samrocketman commented:
> I'm not okay with comments that don't get a reply; your time is very valuable, but equally so is time of the persons that make an effort to do a good review your PR. (And we actually have more people creating PRs than that we have reviewers, so good reviews are in short supply. Don't waste them.)

I went through the earlier feedback and am missing what I failed to address.  Can you point it out?

samrocketman commented:
I'll clarify what I mean perfect being the enemy of good.

For example, there's some PRs attempting to work with atmospheric sounds and music in a way that would likely be solved with a syntax like this PR.

The longer it stays open the higher the likelihood there's risk of duplicate effort and merged conflicts (not referring to files like copyright but C++ classes where music is involved).

From a product perspective: it makes sense to get the feature in a mature enough state that merging it doesn't leave the game worse off and enables maturing the feature further for others to use.

I'm not sure I could have communicated that better in this process. I even went so far as to update the pull request description with ES manual wiki-quality examples and RisingLeaf was invited to help address some of the more technical comments (which they did).

samrocketman commented:
I can work on resolving conflicts

petervdmeer commented:
> > I'm not okay with comments that don't get a reply; your time is very valuable, but equally so is time of the persons that make an effort to do a good review your PR. (And we actually have more people creating PRs than that we have reviewers, so good reviews are in short supply. Don't waste them.)
> 
> I went through the earlier feedback and am missing what I failed to address. Can you point it out?

I sent you a direct message.

> It doesn't make sense to resolve conflicts if there's no intent to merge.

The reason that I reviewed this PR is that I DO have the intent to merge. But the PR does need to be ready for merging when I want to press the merge button.

petervdmeer commented:
> I'll clarify what I mean perfect being the enemy of good.
> 
> For example, there's some PRs attempting to work with atmospheric sounds and music in a way that would likely be solved with a syntax like this PR.

Yes, I'm also convinced that this PR has value. That's why I would like it to be merged.

> The longer it stays open the higher the likelihood there's risk of duplicate effort and merged conflicts (not referring to files like copyright but C++ classes where music is involved).

That is indeed two risks that exist.

> From a product perspective: it makes sense to get the feature in a mature enough state that merging it doesn't leave the game worse off and enables maturing the feature further for others to use. In my opinion, this feature reached that goal.

If I see a comment from Risingleaf on 18 Dec 2023 suggesting a change to make the code more fail-safe, and no reply from the PR-author on such comment, then I'm a little bit worried. Why didn't the comment get a follow-up? (I didn't look at the details, so the comment could be bogus, but then I still expect some sort of reply and "Resolve Conversation" action to happen on such comment.)

> I'm not sure I could have communicated that better in this process. I even went so far as to update the pull request description with ES manual wiki-quality examples (for copy-paste) and RisingLeaf was invited as a collaborator to help address some of the more technical comments (which they did).

That is all very helpful to get the PR to a merge-able state. But open review-comments and merge-conflicts also need to be resolved.

samrocketman commented:
I'll make another pass at the outstanding feedback and resolve the conflicts.  If I'm not sure how to resolve the feedback due to my technical inability then I'll mention so as a reply to the comment.

Some unaddressed feedback comes from my not understanding what the original commenter meant but now they're collaborators and at the very least I can ask about what I don't understand.

Look forward to those updates.

samrocketman commented:
ETA for updates is roughly next week; I'm unable to this week.

samrocketman commented on data/_ui/interfaces.txt:
```diff
@@ -456,6 +456,10 @@ interface "load menu"
 	sprite "ui/keys panel"
 		center -65 -20
+	visible if "show music slider"
+	sprite "ui/music volume"
+		center 330 -60
```
This still needs done I think

RisingLeaf replied:
addressed

samrocketman commented on source/Playlist.h:
```diff
@@ -0,0 +1,67 @@
+	void Activate() const;
+	const Track *GetCurrentTrack() const;
+
+	bool MatchingConditions(PlayerInfo &player) const;
```
As of right now `MatchingConditions` works for selecting music initially.

Can you clarify your question?  I don't understand the ask.

eebop replied:
I think what Amazinite means is to change `PlayerInfo &player` to `const PlayerInfo &player`, as the player is not being modified and that is more clear

RisingLeaf replied:
has been made const &

samrocketman commented on source/Track.h:
```diff
@@ -0,0 +1,63 @@
+
+	int wait = 0;
+
+	bool finishedLoading = false;
```
This is to match patterns of other object types in endless sky.  It's not used yet but this feature is not 100% mature and I anticipate it being needed.

This is an initial feature settling on the data file syntax but it is very incomplete as far as what ES needs in general from a music engine.

samrocketman commented on source/Track.cpp:
```diff
@@ -0,0 +1,103 @@
+	}
+	name = node.Token(1);
+
+	for(const DataNode &child : node)
```
It looks like this was addressed by RisingLeaf.

samrocketman commented on source/Track.cpp:
```diff
@@ -0,0 +1,103 @@
+	for(const DataNode &child : node)
+	{
+		if(child.Token(0) == "volume" && child.Size() >= 2)
+			volumeModifier = child.Value(1);
```
I'll do value checking here and throw a warning.  This is now a TODO thanks.

RisingLeaf replied:
this is clamped

samrocketman commented on source/PreferencesPanel.cpp:
```diff
@@ -114,6 +114,9 @@ void PreferencesPanel::Draw()
 	Information info;
 	info.SetBar("volume", Audio::Volume());
+	info.SetBar("music volume", Audio::MusicVolume());
+	if(!GameData::HasPlaylists())
+		info.SetCondition("show music slider");
```
I think this has been addressed with reasoning.  If you disagree feel free to reply again; for now I'm resolving.

RisingLeaf commented on source/Audio.cpp:
```diff
@@ -255,6 +292,114 @@ void Audio::Update(const Point &listenerPosition)
+			currentPlaylistTrack = currentPlaylist->GetCurrentTrack();
+			if(currentPlaylistTrack)
+			{
+				musicVolumeModifier = currentPlaylistTrack->GetVolumeModifier();
```
changed it, I like it more with the extra variable

RisingLeaf commented:
merge conflicts have been resolved

TomGoodIdea commented on source/Audio.cpp:
```diff
@@ -110,6 +117,17 @@ namespace {
+	const Track *currentPlaylistTrack = nullptr;
+	Track::GameState oldState = Track::GameState::IDLE;
+
+	std::chrono::time_point<std::chrono::high_resolution_clock> onEnd;
```
```suggestion
	chrono::time_point<chrono::high_resolution_clock> onEnd;
```
`std::`s are unnecessary

TomGoodIdea commented on source/Audio.cpp:
```diff
@@ -228,6 +247,24 @@ void Audio::SetVolume(double level)
+// Set the volume (to a value between 0 and 1).
+void Audio::SetMusicVolume(double level)
+{
+	musicVolume = min(1., max(0., level));
```
now that we have C++ 17 support, we can use `std::clamp`
```suggestion
	musicVolume = clamp(level, 0., 1.);
```

TomGoodIdea commented on source/Audio.cpp:
```diff
@@ -255,6 +292,112 @@ void Audio::Update(const Point &listenerPosition)
+			return;
+		}
+	}
+	else
+		if(player.GetSystem() && !player.GetSystem()->MusicName().empty())
+		{
+			PlayMusic(player.GetSystem()->MusicName());
+			return;
+		}
```
```suggestion
	else if(player.GetSystem() && !player.GetSystem()->MusicName().empty())
	{
		PlayMusic(player.GetSystem()->MusicName());
		return;
	}
```

TomGoodIdea commented on source/Audio.cpp:
```diff
@@ -255,6 +292,112 @@ void Audio::Update(const Point &listenerPosition)
+		{
+			isWaiting = true;
+			currentWait = currentPlaylistTrack->Wait();
+			onEnd = std::chrono::high_resolution_clock::now();
```
```suggestion
			onEnd = chrono::high_resolution_clock::now();
```

TomGoodIdea commented on source/Audio.cpp:
```diff
@@ -255,6 +292,112 @@ void Audio::Update(const Point &listenerPosition)
+		}
+		if(isWaiting)
+		{
+			std::chrono::duration<double> elapsed = std::chrono::high_resolution_clock::now() - onEnd;
```
```suggestion
			chrono::duration<double> elapsed = chrono::high_resolution_clock::now() - onEnd;
```

TomGoodIdea commented on source/Playlist.cpp:
```diff
@@ -0,0 +1,139 @@
+
+namespace {
+	const Track *currentTrack = nullptr;
+	const std::set<std::string> PROGRESSION_STYLES = {"random", "linear", "pick"};
```
Maybe it would be better to change this to an enum and just resolve the strings at load?

RisingLeaf replied:
done

TomGoodIdea commented on source/Playlist.cpp:
```diff
@@ -0,0 +1,139 @@
+{
+	// Linear should always get the first track in the list when activating.
+	if(progressionStyle == "linear" && tracks.size() > 0)
+		currentTrack = *tracks.begin();
```
just implement `front()` for `WeightedList`s

RisingLeaf replied:
I am not sure I understand?

RisingLeaf replied:
implemented front

TomGoodIdea commented on source/PreferencesPanel.cpp:
```diff
@@ -140,6 +140,8 @@ void PreferencesPanel::Draw()
 	Information info;
 	info.SetBar("volume", Audio::Volume());
+	info.SetBar("music volume", Audio::MusicVolume());
+	info.SetCondition("show music slider");
```
What's the point of this condition if it's always set?

RisingLeaf replied:
removed

samrocketman replied:
The point of the condition is to hide the slider if there is no music.  The original thought is why show a slider that does nothing from a gamer perspective.

Does the music slider remain hidden if there is no music?

RisingLeaf replied:
-> no as said by @Amazinite here: https://github.com/endless-sky/endless-sky/pull/8491#discussion_r1438920488

samrocketman replied:
I replied to them as well.  I think it will be odd to have a slider that does nothing in the UI but if you both are adamant about it I don't plan to revert it.

Currently, (and I don't think there will be any time soon) content creators have no plans to integrate music. I shopped the idea often when I had the plugin working and there wasn't any interest from the vanilla content creators.

Meaning vanilla is going to have a volume slider that doesn't do anything.

TomGoodIdea replied:
Doesn't the new slider also control ambient system music and the background noises that play when you land on a station?

samrocketman replied:
It's not the same.  It controls the playlist tracks.  As far as I know the sounds slider still controls all of the old definitions of sound.

TomGoodIdea commented on source/Track.cpp:
```diff
@@ -0,0 +1,97 @@
+		bool hasValue = child.Size() >= 2;
+		if(key == "volume" && hasValue)
+			volumeModifier = clamp<double>(child.Value(1), -1., 1.);
+		else if(child.Token(0) == "idle" && child.Size() >= 2)
+			idleTitle = child.Token(1);
+		else if(child.Token(0) == "combat" && child.Size() >= 2)
+			combatTitle = child.Token(1);
+		else if(child.Token(0) == "landed" && child.Size() >= 2)
+			landedTitle = child.Token(1);
+		else if(child.Token(0) == "wait" && child.Size() >= 2)
```
```suggestion
		else if(key == "idle" && hasValue)
			idleTitle = child.Token(1);
		else if(key == "combat" && hasValue)
			combatTitle = child.Token(1);
		else if(key == "landed" && hasValue)
			landedTitle = child.Token(1);
		else if(key == "wait" && hasValue)
```

TomGoodIdea commented on source/Track.cpp:
```diff
@@ -0,0 +1,97 @@
+
+
+
+const std::string &Track::Name() const
```
```suggestion
const string &Track::Name() const
```

TomGoodIdea commented on source/Track.cpp:
```diff
@@ -0,0 +1,97 @@
+
+
+
+const std::string &Track::GetTitle(GameState state) const
```
```suggestion
const string &Track::GetTitle(GameState state) const
```

TomGoodIdea commented on source/Track.cpp:
```diff
@@ -0,0 +1,97 @@
+
+
+
+const double Track::GetVolumeModifier() const
```
primitives don't need to be const
```suggestion
double Track::GetVolumeModifier() const
```

samrocketman replied:
Can you elaborate on why primitives don’t need const?  I’m curious

TomGoodIdea replied:
https://github.com/endless-sky/endless-sky/pull/8621#discussion_r1598812992

TomGoodIdea commented on source/Track.cpp:
```diff
@@ -0,0 +1,97 @@
+
+
+
+const int Track::Wait() const
```
same
```suggestion
int Track::Wait() const
```

TomGoodIdea commented on source/Track.h:
```diff
@@ -0,0 +1,64 @@
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef TRACK_H
+#define TRACK_H
```
```suggestion
#ifndef TRACK_H_
#define TRACK_H_
```

samrocketman replied:
What’s the purpose of the trailing underscore?

TomGoodIdea replied:
doesn't really matter, but in every other header it's `CLASS_NAME_H_` with an underscore at the end

TomGoodIdea commented on source/Track.h:
```diff
@@ -0,0 +1,64 @@
+	const std::string &Name() const;
+
+	const std::string &GetTitle(GameState state) const;
+	const double GetVolumeModifier() const;
+	const int Wait() const;
```
```suggestion
	double GetVolumeModifier() const;
	int Wait() const;
```

TomGoodIdea commented on source/UniverseObjects.h:
```diff
@@ -118,9 +120,11 @@ class UniverseObjects {
 	Set<Ship> ships;
 	Set<System> systems;
 	Set<Test> tests;
+	Set<Track> tracks;
 	Set<TestData> testDataSets;
```
sort alphabetically
```suggestion
	Set<TestData> testDataSets;
	Set<Track> tracks;
```

TomGoodIdea reviewed (commented):
mostly style fixes for now

RisingLeaf commented on source/Audio.cpp:
```diff
@@ -255,6 +292,111 @@ void Audio::Update(const Point &listenerPosition)
+	if(!isInitialized)
+		return;
+
+	// Music defined by planet
```
```suggestion
	// Music defined by planet or system.
```

RisingLeaf commented on source/Track.h:
```diff
@@ -0,0 +1,64 @@
+// Class to store a track of music that can be used in a playlist.
+class Track {
+public:
+	enum class GameState : int_fast8_t
```
```suggestion
	enum GameState
```

RisingLeaf commented on copyright:
```diff
@@ -3554,6 +3554,13 @@ Copyright: RisingLeaf (https://github.com/RisingLeaf)
 Comment: Derived from _menu/haze-133 (no copyright given).
 
+Files:
+ images/projectile/ionic?turret*
+ images/star/proto-planetary-disk*
```
```suggestion
```

tibetiroka commented on source/Audio.cpp:
```diff
@@ -255,6 +292,111 @@ void Audio::Update(const Point &listenerPosition)
+			currentWait = currentPlaylistTrack->Wait();
+			onEnd = chrono::high_resolution_clock::now();
+		}
+		if(isWaiting)
```
```suggestion
		else if(isWaiting)
```

tibetiroka commented on source/Audio.cpp:
```diff
@@ -255,6 +292,111 @@ void Audio::Update(const Point &listenerPosition)
+		}
+	}
+
+	// If the current playlists conditions are not matching anymore, search a new one.
```
```suggestion
	// If the current playlist's conditions are not matching anymore, search for a new playlist.
```

tibetiroka commented on source/Audio.cpp:
```diff
@@ -255,6 +292,111 @@ void Audio::Update(const Point &listenerPosition)
+		currentPlaylist->MatchingConditions(player) : false;
+	// The track has to be updated if the current track is finished or the playlist is not matching
+	// anymore.
+	if(currentTrack->IsFinished() || !currentPlaylistValid)
```
If I'm reading this code correctly, this could just be split up into two consecutive if blocks?

RisingLeaf replied:
done

tibetiroka commented on source/Audio.cpp:
```diff
@@ -110,6 +117,17 @@ namespace {
 	int musicFade = 0;
 	vector<int16_t> fadeBuffer;
+	double musicVolume = 1.;
+	double playlistVolumeModifier = 0.;
```
```suggestion
	// The volume modifier of the current track.
	double playlistVolumeModifier = 0.;
```
Should it be called `trackVolumeModifier`, since it's actually defined per track?

RisingLeaf replied:
ye probably

tibetiroka commented on source/Playlist.cpp:
```diff
@@ -0,0 +1,142 @@
+			toPlay.Load(child);
+		else if(key == "location")
+			location.Load(child);
+		else if(key == "priority" && hasValue)
+			priority = max<unsigned>(0, child.Value(1));
+		else if(key == "weight" && hasValue)
+			weight = max<unsigned>(1, child.Value(1));
```
These should print warnings if the user-defined values get ignored, as it's not intuitive that a weight of 0.5 doesn't function, or that a negative priority doesn't get less priority than a default-valued playlist.

Are these even necessary? I think checking for a positive weight should be enough.

RisingLeaf replied:
now checking for negative values -> warning

tibetiroka commented on source/Playlist.cpp:
```diff
@@ -0,0 +1,142 @@
+{
+	switch(progressionStyle)
+	{
+	case ProgressionStyle::LINEAR:
+		{
+			const Track *tmpTrack = currentTrack;
+			auto it = find(tracks.begin(), tracks.end(), tmpTrack);
+			++it;
+			if(it == tracks.end())
+				it = tracks.begin();
+			currentTrack = *it;
+			return tmpTrack;
+		}
+	case ProgressionStyle::PICK:
+		return currentTrack;
+	case ProgressionStyle::RANDOM:
+	default:
+		return tracks.Get();
```
This should be indented more, unless GitHub is messing it up for me.

RisingLeaf replied:
done

tibetiroka commented on source/Playlist.cpp:
```diff
@@ -0,0 +1,142 @@
+
+
+
+const Track *Playlist::GetCurrentTrack() const
```
Okay, so this isn't a simple getter for `currentTrack`? That should be documented.

Also, it does seem like it would be better if each playlist kept track of its current track. That should avoid most of the complications here, though it may require a `Deactivate()` function.

RisingLeaf replied:
there is max one playlist active, I dont think thats necessary

RisingLeaf replied:
also this way the function can be const

RisingLeaf replied:
also renamed

tibetiroka replied:
At the very least, this should have a comment. It's not in line with our general function naming scheme.

RisingLeaf replied:
it has a comment

tibetiroka commented on source/Playlist.h:
```diff
@@ -0,0 +1,75 @@
+	void Activate() const;
+	const Track *GetCurrentTrack() const;
+
+	bool MatchingConditions(const PlayerInfo &player) const;
```
```suggestion
	bool MatchesConditions(const PlayerInfo &player) const;
```

tibetiroka commented on source/Track.cpp:
```diff
@@ -0,0 +1,97 @@
+{
+	if(!name.empty())
+	{
+		node.PrintTrace("Error: Duplicate definition of track:");
```
```suggestion
		node.PrintTrace("Skipping duplicate definition of track:");
```

RisingLeaf commented:
https://github.com/endless-sky/endless-sky-wiki/pull/32

tibetiroka commented on source/Audio.cpp:
```diff
@@ -110,6 +117,18 @@ namespace {
 	shared_ptr<Music> previousTrack;
 	int musicFade = 0;
 	vector<int16_t> fadeBuffer;
+	double musicVolume = 1.;
```
This should default to the same value as the sound volume, `.125`.

tibetiroka reviewed (commented):
Based on some settings, it looks like the volume sliders are completely broken. Changing either slider changes the music volume.

tibetiroka commented on source/Track.cpp:
```diff
@@ -0,0 +1,97 @@
+		bool hasValue = child.Size() >= 2;
+		if(key == "volume" && hasValue)
+			volumeModifier = clamp<double>(child.Value(1), -1., 1.);
+		else if(key == "idle" && hasValue)
+			idleTitle = child.Token(1);
+		else if(key == "combat" && hasValue)
+			combatTitle = child.Token(1);
+		else if(key == "landed" && hasValue)
+			landedTitle = child.Token(1);
```
I think `combat` and `landed` should default to `idle` if they aren't specified.

tibetiroka commented on source/Playlist.cpp:
```diff
@@ -55,9 +55,17 @@ void Playlist::Load(const DataNode &node)
+		{
 			priority = max<unsigned>(0, child.Value(1));
+			if(child.Value(1) < 0)
+				child.PrintTrace("Warning: Negative weight sepcified for priority.");
```
```suggestion
				child.PrintTrace("Warning: Ignoring negative priority:");
```
~~Just support negative priorities~~

tibetiroka commented on source/Playlist.cpp:
```diff
@@ -55,9 +55,17 @@ void Playlist::Load(const DataNode &node)
+		{
 			weight = max<unsigned>(1, child.Value(1));
+			if(child.Value(1) < 0)
+				child.PrintTrace("Warning: Negative weight sepcified for priority.");
```
```suggestion
				child.PrintTrace("Warning: Ignoring negative weight");
```
This shouldn't allow 0 either. It should allow fractional values, though, as that seems to be an arbitrary restriction.

RisingLeaf replied:
WeightedList uses only uint iirc

tibetiroka replied:
Ah.

RisingLeaf commented:
> Based on some settings, it looks like the volume sliders are completely broken. Changing either slider changes the music volume.

Yep one is master one is music, standard

tibetiroka commented:
Okay, I didn't realize that.

Unrelated, but it seems like you can only play one track before something in the code breaks.

RisingLeaf commented:
Like a crash?

tibetiroka commented:
No crash, it just doesn't play the next track.

tibetiroka commented:
Yes. Though I don't get music in the load/save menu every the time either.

tibetiroka commented:
It works now.

tibetiroka reviewed (changes requested):
Based on a discussion with the other devs, it looks like the current implementation is not headed the way that would be most useful in the game. Since music is mostly for ambiance, having playlists isn't necessary. In fact, based on my previous discussion with @RisingLeaf, it seems to just create more problems than it solves.

So instead of the PR dealing with playlists of tracks, we feel we should only have individual tracks that have their own `to play` conditions and `location` filters. For now, the entire syntax would be reduced to just:
```
track <name>
	music <name>
	["to play"]
		<conditions>
	[location]
		<location filter>
```
(The `volume` and `wait` parameters could also be included, if that makes things easier.)

In the future, this could be expanded to have idle, combat or landed tracks, but let's keep things simple for now. Systems and planets would retain the ability to specify _tracks_ that should be played there, overriding the default selections.

Essentially, this means dropping playlists, weights, priorities, track variations and progression styles. Some of these may be added in the future, but that should be done in follow-up PRs.

samrocketman commented:
The original intent of the music engine and playlist feature was to be event-based. In general, I think it wouldn't be in-line with the original vision of the music engine feature to not have playlists and the ability to trigger tracks conversationally.

For example, imagine tracks triggering where there's bustling bar noise when a conversation involves a bar.  I can't imagine why the current devs wouldn't want this considering the efforts put forth for it but I understand it's my opinion.

samrocketman commented:
The PR was opened and meant to be merged and iterated on. Feedback like there being no menu music is feedback for the current poor support in ES.  i.e. it doesn't have much of a music engine so things like that wouldn't be it.  It's not really feedback for this PR and that seemed to be a confusing point.

The point is this was meant to be a starting point to build out a lexicon of syntax for triggering music. And then building on it to mature the engine support (like adding the ability for menu music).

petervdmeer commented:
As discussed on discord (with Rising, tibetiroka and me doing most of the talking):
- Whatever syntax we choose should include seamless transitions.
  - We might also have songs that seamless transfer into other songs, but not back.
- Playlists can be usefull to avoid repeating the same conditions over and over again, if we have multiple songs all with the same conditions.
- The following features are also desired, and could be implemented without playlists:
  - Playing songs "in order" can be done by specifying which song typically follows another song.
    - We could create an automatic condition that tells us which song was playing (and use that for `follows`), but it is probably more complex than just using a separate keyword for this. (So the keyword is preferred.)
  - Playing songs repeatedly (the pick-keyword in this PR) can also be handled by a repeat keyword on a track.
  - Choosing songs "at random" would be the default in whatever we do.
- We should be able to switch music on any event (not only on combat/idle/landing).
- Priorities can be usefull to avoid a lot of additional conditions.
- Tracks can be identified by filenames, they don't need additional names on top of that.
- Ambient sounds are typically just 1 track that repeats over and over again (or a set of seamlessly transitioning tracks).

An example of a syntax that could also support all features (except playlists to avoid repeating the same conditions over and over again):

```html
track <filename>
	"to play"
		<conditions for playing this song>
	priority <nr to indicate if this song gets chosen over other applicable songs>
	location
		<location filter>
	"transfer from"
		<filename of song to seamlessly transfer from>
	follows
		<filename of other songs which this song typically plays after>
	[repeat]
```

Rising wants to make an updated proposal somewhere next week including the features discussed, and including playlists, so let's continue from that updated proposal once it is available.

RisingLeaf commented:
Better late than never, here is my proposal to meet halfway:
```html
playlist <name>
        priority <nr of priority, overrides the playlist>
        "to play"
                <conditions>
        location
                <location filter>

track <name>
        music <name of the file>
        playlist <name of the playlist to load conditions from>
        priority <nr of priority, overrides the playlist>
        "to play"
                <conditions that add to those of the playlist>
        location
                <location filter that adds to the one of the playlist>
        "transfer from"
                <filenames of songs to seamlessly transfer from>
        follows
                <filenames of other songs which this song typically plays after>
        [repeat]

RisingLeaf commented:
Most of the time conditions and locations should be done via the playlist to avoid clutter and repetition

tibetiroka commented:
Closing as the current maintainer of the PR is no longer active.

samrocketman commented:
Reason for closing is in this thread.  https://discord.com/channels/251118043411775489/989659299645100112/1274400760947408994

The reality is we reached an impasse and can’t agree on the feature.  I guess you will decide what to do with it without my feedback.

tibetiroka commented:
As evidence by Leaf's comment above, we did reach an agreement. He just isn't maintaining the PR anymore.

We discussed it in the core VC as previously agreed. Leaf, Peter and I showed up; you didn't. The three of us reached an agreement, and Leaf also posted it here; you didn't even comment on Leaf's proposal here to show that you don't agree with it.

It's still up for discussion, if you have any  feedback. But if you don't intend to take over maintenance of the PR again, I don't see any reason to reopen it without a maintainer.

samrocketman commented:
I see; so the fact that I reopened this PR and added hundreds of lines to make it stable doesn't count as a maintainer.

Nobody responded to Leaf's comment from the core team; 

**why would I do work on a PR when there's open disagreement from that thread and no conclusion posted in this PR?**

Me not disagreeing with Leaf's proposal doesn't mean you or any other dev "agrees" with it.  I'm tired of working on something that doesn't have quorum.

@tibetiroka your closing of this PR is more of an indication of disagreement here than any feedback.  There's nothing wrong with me taking a wait and see approach to see responses.

samrocketman commented:
I don't plan to work on this further; I don't like how this feature was handled start to finish and am calling it quits rather than argue further.

I'm strongly considering pulling out of the project all-together which I'm sure won't affect the dev team much since I haven't opened recent PRs but that was more of a change of approach waiting on open PRs to be finished before opening more.
